dev/AppLifecycle/AppInstance.cpp

```diff
@@ -0,0 +1,421 @@
+        auto name = wil::str_printf<std::wstring>(c_requestPacketNameFormat, m_processName.c_str(), idString.get());
+
+        RedirectionRequest request;
+        request.Open(name);
+        request.MarshalArguments(args);
+
+        std::wstring eventName = name + c_activatedEventNameSuffix;
+        wil::unique_event cleanupEvent;
+        cleanupEvent.create(wil::EventOptions::ManualReset, eventName.c_str());
+
+        // Enqueue the request and signal the other instance.
+        EnqueueRedirectionRequestId(id);
+        m_innerActivated.SetEvent();
+
+        // Wait for the other instance to open the memory mapped file before exiting and cleaning our interest in it.
+        cleanupEvent.wait();
+        co_return;
+    }
+
+    IAsyncAction AppInstance::RedirectActivationToAsync(AppLifecycle::AppActivationArguments const& args)
+    {
+        if (!m_isCurrent)
+        {
+            co_await QueueRequest(args);
+        }
+    }
+
+    Microsoft::Windows::AppLifecycle::AppInstance AppInstance::GetCurrent()
+    {
+        auto initInstance = []
+        {
+            s_current = winrt::make_self<AppInstance>(GetCurrentProcessId());
+        };
+
+        wil::init_once(s_initOnce, initInstance);
+
+        return s_current.as<Microsoft::Windows::AppLifecycle::AppInstance>();
+    }
+
+    IVector<Microsoft::Windows::AppLifecycle::AppInstance> AppInstance::GetInstances()
+    {
+        IVector<Microsoft::Windows::AppLifecycle::AppInstance> instances{ winrt::single_threaded_vector<Microsoft::Windows::AppLifecycle::AppInstance>() };
+
+        // Grab the list of processes while under the lock, and then drop it since we'll be calling out to other code.
+        std::list<DWORD> pids;
+        {
+            auto releaseOnExit = s_current->m_dataMutex.acquire();
+
+            for (uint32_t index = 0; index < s_current->m_instances.Size(); index++)
+            {
+                auto pid = s_current->m_instances[index];
+                if (pid == 0)
+                {
+                    continue;
+                }
+
+                pids.push_back(pid);
+            }
+        }
+
+        // Create the associated AppInstance objects while removing orphaned entries we find.
+        for (const auto& pid : pids)
+        {
+            if (GetCurrentProcessId() == pid)
+            {
+                instances.Append(AppInstance::GetCurrent());
+            }
+            else
+            {
+                wil::unique_handle process(::OpenProcess(PROCESS_QUERY_LIMITED_INFORMATION, FALSE, pid));
+                if (process != nullptr)
+                {
+                    instances.Append(make<AppInstance>(pid));
+                }
+                else
+                {
+                    // Remove orphan.
+                    auto releaseOnExit = s_current->m_dataMutex.acquire();
+                    for (uint32_t index = 0; index < s_current->m_instances.Size(); index++)
```

Comment:
SharedProcessList::Remove already seems to be checking whether pid is present in the list or not. Maybe this loop can be avoided?